### PR TITLE
Pv2 1381 populate v1 payments tables

### PR DESCRIPTION
**Pull request by carlbm**

### CODE REVIEW

File: src/SFA.DAS.Payments.ProviderPayments.Application/Repositories/LegacyPaymentsRepository.cs [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+
+namespace SFA.DAS.Payments.ProviderPayments.Application.Repositories
+{
+    public interface ILegacyPaymentsRepository
+    {
+        Task WritePaymentInformation(IEnumerable<LegacyPaymentModel> payments,
+            IEnumerable<LegacyRequiredPaymentModel> requiredPayments);
+    }
+
+    public class LegacyPaymentsRepository
```
Comment (samuel-dev-dba) [inline src/SFA.DAS.Payments.ProviderPayments.Application/Repositories/LegacyPaymentsRepository.cs]: We don't seem to be implementing the interface - ILegacyPaymentsRepository

File: src/SFA.DAS.Payments.ProviderPayments.Application/Repositories/LegacyPaymentsRepository.cs [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+    {
+        private readonly string connectionString;
+
+        public LegacyPaymentsRepository(IConfigurationHelper configurationHelper)
```
Comment (samuel-dev-dba) [inline src/SFA.DAS.Payments.ProviderPayments.Application/Repositories/LegacyPaymentsRepository.cs]: Can we inject an IBulk Writer Config

File: src/SFA.DAS.Payments.ProviderPayments.Application/Services/PaymentExportService.cs [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+using SFA.DAS.Payments.ProviderPayments.Domain.Services;
+
+namespace SFA.DAS.Payments.ProviderPayments.Application.Services
+{
+    public interface IPaymentExportService
+    {
+        Task PerformExportPaymentsToV1(CollectionPeriod collectionPeriod);
+    }
+
+    class PaymentExportService : IPaymentExportService
```
Comment (samuel-dev-dba) [inline src/SFA.DAS.Payments.ProviderPayments.Application/Services/PaymentExportService.cs]: Do we have any unit test for this class

File: src/SFA.DAS.Payments.ProviderPayments.Application/Services/PaymentExportService.cs [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+            IProviderPaymentsRepository providerPaymentsRepository, 
+            IPaymentLogger logger, 
+            IPaymentMapper paymentMapper,
+            IConfigurationHelper configurationHelper)
+        {
+            this.providerPaymentsRepository = providerPaymentsRepository;
+            this.logger = logger ?? throw new ArgumentNullException(nameof(logger));
+            this.legacyPaymentsRepository = legacyPaymentsRepository ?? throw new ArgumentNullException(nameof(legacyPaymentsRepository));
+            this.paymentExportProgressCache = paymentExportProgressCache ?? throw new ArgumentNullException(nameof(paymentExportProgressCache));
+            exportBufferSize = configurationHelper.GetSettingOrDefault("ExportBatchSize", 10000);
```
Comment (samuel-dev-dba) [inline src/SFA.DAS.Payments.ProviderPayments.Application/Services/PaymentExportService.cs]: Can this be renamed to exportBatchSize

File: src/SFA.DAS.Payments.ProviderPayments.Application/Services/PaymentExportService.cs [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+            this.logger = logger;
+            this.paymentMapper = paymentMapper ?? throw new ArgumentNullException(nameof(paymentMapper));
+        }
+
+        public async Task PerformExportPaymentsToV1(CollectionPeriod collectionPeriod)
+        {
+            logger.LogVerbose($"Started V1 payments export for collection period {collectionPeriod}");
+
+            while (true)
+            {
```
Comment (samuel-dev-dba) [inline src/SFA.DAS.Payments.ProviderPayments.Application/Services/PaymentExportService.cs]: I think we need to handle Cancellation token here so it throws an exception if the token is cancelled

File: src/SFA.DAS.Payments.ProviderPayments.Application/Services/PaymentExportService.cs [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+            this.providerPaymentsRepository = providerPaymentsRepository;
+            this.logger = logger ?? throw new ArgumentNullException(nameof(logger));
+            this.legacyPaymentsRepository = legacyPaymentsRepository ?? throw new ArgumentNullException(nameof(legacyPaymentsRepository));
+            this.paymentExportProgressCache = paymentExportProgressCache ?? throw new ArgumentNullException(nameof(paymentExportProgressCache));
+            exportBufferSize = configurationHelper.GetSettingOrDefault("ExportBatchSize", 10000);
+            this.logger = logger;
+            this.paymentMapper = paymentMapper ?? throw new ArgumentNullException(nameof(paymentMapper));
+        }
+
+        public async Task PerformExportPaymentsToV1(CollectionPeriod collectionPeriod)
```
Comment (samuel-dev-dba) [inline src/SFA.DAS.Payments.ProviderPayments.Application/Services/PaymentExportService.cs]: We need to pass in a cancellation token

File: src/SFA.DAS.Payments.ProviderPayments.Application/Services/PaymentExportService.cs [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+        {
+            logger.LogVerbose($"Started V1 payments export for collection period {collectionPeriod}");
+
+            while (true)
+            {
+                var page = await paymentExportProgressCache.GetPage(collectionPeriod.AcademicYear, collectionPeriod.Period);
```
Comment (samuel-dev-dba) [inline src/SFA.DAS.Payments.ProviderPayments.Application/Services/PaymentExportService.cs]: This code block can do with a try-catch  and log error

File: src/SFA.DAS.Payments.ProviderPayments.Application/Services/PaymentExportService.cs (hunk 5 above)
Comment (carlbm) [reply src/SFA.DAS.Payments.ProviderPayments.Application/Services/PaymentExportService.cs]: It's coming from a handler at present which doesn't have a cancellation token

File: src/SFA.DAS.Payments.ProviderPayments.Application/Services/PaymentExportService.cs (hunk 6 above)
Comment (carlbm) [reply src/SFA.DAS.Payments.ProviderPayments.Application/Services/PaymentExportService.cs]: It's coming from a handler at present which doesn't have a cancellation token

**Comment by abroodie** [comment]
The story for this work clearly states that this should be handled in a new service fabric service and that the batching should be per collection not per provider.

File: src/SFA.DAS.Payments.ProviderPayments.ServiceFabric/ApplicationParameters/Cloud.xml [hunk 8]
```diff
@@ -16,5 +16,7 @@
     <Parameter Name="ProviderPaymentsService_EndpointName" Value="__ProviderPaymentsService_EndpointName__" />
     <Parameter Name="ApplicationInsightsInstrumentationKey" Value="__ApplicationInsights_InstrumentationKey__" />
     <Parameter Name="BatchIntervalInSeconds" Value="__BatchIntervalInSeconds__" />
+    <Parameter Name="DasPeriodEndConnectionString" Value="__DasPeriodEndConnectionString__" />
+    <Parameter Name="ExportBatchSize" Value="__ExportBatchSize__" />
```
Comment (abroodie) [inline src/SFA.DAS.Payments.ProviderPayments.ServiceFabric/ApplicationParameters/Cloud.xml]: The vsts variable should have a more explicit name, e.g. LegacyPaymentsBatchSize.

File: src/SFA.DAS.Payments.ProviderPayments.Application/Services/PaymentExportService.cs [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+                    .WritePaymentInformation(result.payments, result.requiredPayments)
+                    .ConfigureAwait(false);
+
+                logger.LogVerbose($"Completed write for page: {page} for collection period: {collectionPeriod}");
+
+                await paymentExportProgressCache
+                    .IncrementPage(collectionPeriod.AcademicYear, collectionPeriod.Period)
+                    .ConfigureAwait(false);
+            }
+
```
Comment (abroodie) [inline src/SFA.DAS.Payments.ProviderPayments.Application/Services/PaymentExportService.cs]: The code should be writing to the period end table too or else the MA finance/accounts service won't know the payments are available.

File: src/SFA.DAS.Payments.ProviderPayments.Domain/Services/PaymentMapper.cs [hunk 10]
```diff
@@ -0,0 +1,110 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using SFA.DAS.Payments.ProviderPayments.Application.Data;
+using SFA.DAS.Payments.ProviderPayments.Model.V1;
+
+namespace SFA.DAS.Payments.ProviderPayments.Domain.Services
+{
+    public interface IPaymentMapper
```
Comment (abroodie) [inline src/SFA.DAS.Payments.ProviderPayments.Domain/Services/PaymentMapper.cs]: Should also be mapping data for the earnings table too.

File: src/SFA.DAS.Payments.ProviderPayments.ProviderPaymentsService/Handlers/ProcessPaymentsToV1Handler.cs [hunk 11]
```diff
@@ -0,0 +1,17 @@
+﻿using System.Threading.Tasks;
+using NServiceBus;
+using SFA.DAS.Payments.ProviderPayments.Application.Services;
+using SFA.DAS.Payments.ProviderPayments.Messages.Internal.Commands;
+
+namespace SFA.DAS.Payments.ProviderPayments.ProviderPaymentsService.Handlers
+{
+    public class ProcessPaymentsToV1Handler : IHandleMessages<ProcessProviderMonthEndCommand>
```
[the remark this answers is not in the document]
Comment (carlbm) [reply src/SFA.DAS.Payments.ProviderPayments.ProviderPaymentsService/Handlers/ProcessPaymentsToV1Handler.cs]: Changed to PeriodEndStoppedEvent

File: src/SFA.DAS.Payments.ProviderPayments.Application/Services/PaymentExportService.cs (hunk 9 above)
Comment (carlbm) [reply src/SFA.DAS.Payments.ProviderPayments.Application/Services/PaymentExportService.cs]: Added export to earnings and Payments.Periods

File: src/SFA.DAS.Payments.ProviderPayments.Domain/Services/PaymentMapper.cs (hunk 10 above)
Comment (carlbm) [reply src/SFA.DAS.Payments.ProviderPayments.Domain/Services/PaymentMapper.cs]: Changed

File: src/SFA.DAS.Payments.ProviderPayments.ServiceFabric/ApplicationParameters/Cloud.xml (hunk 8 above)
Comment (carlbm) [reply src/SFA.DAS.Payments.ProviderPayments.ServiceFabric/ApplicationParameters/Cloud.xml]: Changed

File: src/SFA.DAS.Payments.ProviderPayments.Application/Repositories/LegacyPaymentsRepository.cs (hunk 1 above)
Comment (carlbm) [reply src/SFA.DAS.Payments.ProviderPayments.Application/Repositories/LegacyPaymentsRepository.cs]: We are now

File: src/SFA.DAS.Payments.ProviderPayments.LegacyExportService/Handlers/ProcessTriggerMonthCompleteHandler.cs [hunk 12]
```diff
@@ -0,0 +1,17 @@
...
+
+namespace SFA.DAS.Payments.ProviderPayments.LegacyExportService.Handlers
+{
+    class ProcessTriggerMonthCompleteHandler : IHandleMessages<ProcessTriggerMonthComplete>
+    {
+        public IPaymentExportService PaymentExportService { get; set; }
+
+        public async Task Handle(ProcessTriggerMonthComplete message, IMessageHandlerContext context)
+        {
+            await PaymentExportService.PerformMonthEndTrigger(message.CollectionPeriod).ConfigureAwait(false);
```
Comment (samuel-dev-dba) [inline src/SFA.DAS.Payments.ProviderPayments.LegacyExportService/Handlers/ProcessTriggerMonthCompleteHandler.cs]: Can we increase NService Bus timeout here to 5 or 10min so we can be sure there is enough time for payments to be copied successfully?

File: src/SFA.DAS.Payments.ProviderPayments.Application/Repositories/ProviderPaymentsRepository.cs [hunk 13]
```diff
@@ -29,6 +30,31 @@ public ProviderPaymentsRepository(IPaymentsDataContext paymentsDataContext)
                 .ToListAsync(cancellationToken);
         }
 
+        public List<PaymentModelWithRequiredPaymentId> GetMonthEndPayments(CollectionPeriod collectionPeriod, int pageSize,
+            int page)
+        {
+            var payments = paymentsDataContext.PaymentsWithRequiredPayments.FromSql($@"
```
[the remark this answers is not in the document]
Comment (carlbm) [reply src/SFA.DAS.Payments.ProviderPayments.Application/Repositories/ProviderPaymentsRepository.cs]: As discussed leaving inline SQL (multiple tables that would need to be configured in EF)

File: src/SFA.DAS.Payments.ProviderPayments.LegacyExportService/Handlers/ProcessTriggerMonthCompleteHandler.cs (hunk 12 above)
Comment (samuel-dev-dba) [reply src/SFA.DAS.Payments.ProviderPayments.LegacyExportService/Handlers/ProcessTriggerMonthCompleteHandler.cs]: I am happy to let this go for now pending test result on big ILR submission and then processing payment in Shadow.

File: src/SFA.DAS.Payments.ProviderPayments.ServiceFabric/ApplicationParameters/Cloud.xml [hunk 14]
```diff
@@ -1,6 +1,9 @@
 ﻿<?xml version="1.0" encoding="utf-8"?>
 <Application xmlns:xsd="http://www.w3.org/2001/XMLSchema" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" Name="fabric:/SFA.DAS.Payments.ProviderPayments.ServiceFabric" xmlns="http://schemas.microsoft.com/2011/01/fabric">
   <Parameters>
+    <Parameter Name="SFA.DAS.Payments.ProviderPayments.LegacyExportService_PartitionCount" Value="1" />
```
Comment (samuel-dev-dba) [inline src/SFA.DAS.Payments.ProviderPayments.ServiceFabric/ApplicationParameters/Cloud.xml]: Need to use the right convention for the config values

File: src/SFA.DAS.Payments.ProviderPayments.Application/Services/PaymentExportService.cs (hunk 3 above)
Comment (carlbm) [reply src/SFA.DAS.Payments.ProviderPayments.Application/Services/PaymentExportService.cs]: Added

File: src/SFA.DAS.Payments.EarningEvents.Application.UnitTests/Mapping/ApprenticeshipContractTypeEarningsEventBuilderTests.cs [hunk 15]
```diff
@@ -0,0 +1,10 @@
+﻿using NUnit.Framework;
+
+namespace SFA.DAS.Payments.EarningEvents.Application.UnitTests.Mapping
+{
+    [TestFixture]
+    public class ApprenticeshipContractTypeEarningsEventBuilderTests
+    {
+
+    }
+}
```
Comment (samuel-dev-dba) [inline src/SFA.DAS.Payments.EarningEvents.Application.UnitTests/Mapping/ApprenticeshipContractTypeEarningsEventBuilderTests.cs]: This class should be removed

Comment (carlbm) [reply src/SFA.DAS.Payments.EarningEvents.Application.UnitTests/Mapping/ApprenticeshipContractTypeEarningsEventBuilderTests.cs]: Removed

File: src/SFA.DAS.Payments.ProviderPayments.ServiceFabric/ApplicationParameters/Cloud.xml (hunk 14 above)
Comment (carlbm) [reply src/SFA.DAS.Payments.ProviderPayments.ServiceFabric/ApplicationParameters/Cloud.xml]: Changed

File: src/SFA.DAS.Payments.ProviderPayments.Application.UnitTests/Services/PaymentExportServiceTests/PerformMonthEndTriggerTests.cs [hunk 16]
```diff
@@ -0,0 +1,138 @@
...
+        [Test]
+        public async Task UsesTheCorrectPage()
+        {
+            paymentExportProgressCache.Setup(x => x.GetPage(testCollectionPeriod.AcademicYear, testCollectionPeriod.Period))
+                .ReturnsAsync(123);
+
+            providerPaymentsRepository.Setup(x => x.GetMonthEndPayments(testCollectionPeriod, 10000, 123))
+                .Returns(testPayments);
+
+            await sut.PerformExportPaymentsAndEarningsToV1(testCollectionPeriod);
```
Comment (samuel-dev-dba) [inline src/SFA.DAS.Payments.ProviderPayments.Application.UnitTests/Services/PaymentExportServiceTests/PerformMonthEndTriggerTests.cs]: Not sure what we are testing here

File: src/SFA.DAS.Payments.ProviderPayments.Application.UnitTests/Services/PaymentExportServiceTests/PerformMonthEndTriggerTests.cs [hunk 17]
```diff
@@ -0,0 +1,138 @@
...
+        }
+
+        [Test]
+        public async Task StopsWhenThereAreZeroPayments()
+        {
+            providerPaymentsRepository.Setup(x => x.GetMonthEndPayments(testCollectionPeriod, 10000, It.IsAny<int>()))
+                .Returns(testPayments);
+
+            await sut.PerformExportPaymentsAndEarningsToV1(testCollectionPeriod);
```
Comment (samuel-dev-dba) [inline src/SFA.DAS.Payments.ProviderPayments.Application.UnitTests/Services/PaymentExportServiceTests/PerformMonthEndTriggerTests.cs]: Not sure what we are testing here

File: src/SFA.DAS.Payments.ProviderPayments.Application.UnitTests/Services/PaymentExportServiceTests/PerformMonthEndTriggerTests.cs (hunk 16 above)
Comment (carlbm) [reply src/SFA.DAS.Payments.ProviderPayments.Application.UnitTests/Services/PaymentExportServiceTests/PerformMonthEndTriggerTests.cs]: That it's using the page returned from the cache...

**Pull request merged**